<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-기계학습(50점)" data-toc-modified-id="1.-기계학습(50점)-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>1. 기계학습(50점)</a></span><ul class="toc-item"><li><span><a href="#1.1-데이터-탐색" data-toc-modified-id="1.1-데이터-탐색-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>1.1 데이터 탐색</a></span><ul class="toc-item"><li><span><a href="#1.1.2.-앞선-두-단계에서-얻은-향후-분석시-고려사항-작성" data-toc-modified-id="1.1.2.-앞선-두-단계에서-얻은-향후-분석시-고려사항-작성-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>1.1.2. 앞선 두 단계에서 얻은 향후 분석시 고려사항 작성</a></span></li></ul></li><li><span><a href="#1.2-클래스-불균형을-처리하시오." data-toc-modified-id="1.2-클래스-불균형을-처리하시오.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>1.2 클래스 불균형을 처리하시오.</a></span><ul class="toc-item"><li><span><a href="#1.2.1.-업-샘플링-과정-설명하고-결과-작성" data-toc-modified-id="1.2.1.-업-샘플링-과정-설명하고-결과-작성-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>1.2.1. 업 샘플링 과정 설명하고 결과 작성</a></span></li><li><span><a href="#1.2.2-언더-샘플링-과정-설명하고-결과-작성" data-toc-modified-id="1.2.2-언더-샘플링-과정-설명하고-결과-작성-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>1.2.2 언더 샘플링 과정 설명하고 결과 작성</a></span></li><li><span><a href="#1.2.3-둘-중-선택하고-이유-설명" data-toc-modified-id="1.2.3-둘-중-선택하고-이유-설명-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>1.2.3 둘 중 선택하고 이유 설명</a></span></li></ul></li><li><span><a href="#1.3.-모델링-하시오" data-toc-modified-id="1.3.-모델링-하시오-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>1.3. 모델링 하시오</a></span><ul class="toc-item"><li><span><a href="#1.3.1.-최소-3개-이상-알고리즘-제시하고-정확도-측면의-모델-1개와-속도-측면의-모델-1개를-꼭-구현(총-2개-이상)" data-toc-modified-id="1.3.1.-최소-3개-이상-알고리즘-제시하고-정확도-측면의-모델-1개와-속도-측면의-모델-1개를-꼭-구현(총-2개-이상)-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>1.3.1. 최소 3개 이상 알고리즘 제시하고 정확도 측면의 모델 1개와 속도 측면의 모델 1개를 꼭 구현(총 2개 이상)</a></span></li><li><span><a href="#1.3.2.-모델-비교하고-결과-설명" data-toc-modified-id="1.3.2.-모델-비교하고-결과-설명-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>1.3.2. 모델 비교하고 결과 설명</a></span></li><li><span><a href="#1.3.3.-속도-개선을-위한-차원-축소-설명하고-수행,-예측-성능과-속도-비교하고-결과-작성" data-toc-modified-id="1.3.3.-속도-개선을-위한-차원-축소-설명하고-수행,-예측-성능과-속도-비교하고-결과-작성-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>1.3.3. 속도 개선을 위한 차원 축소 설명하고 수행, 예측 성능과 속도 비교하고 결과 작성</a></span></li></ul></li></ul></li></ul></div>

## 1. 기계학습(50점)

### 1.1 데이터 탐색

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

df1 = pd.read_csv('../data/diabetes_for_test.csv')
df1


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

당뇨병이 있는 사람은 없는 사람보다 pregnancies의 평균 수치가 높다.

당뇨병이 있는 사람은 없는 사람 보다 Glucose의 평균 수치가 높다.

당뇨병이 있는 사람은 없는 사람 보다 Insulin의 평균 수치가 높다.

당뇨병이 있는 사람은 없는 사람 보다 BMI의 평균 수치가 높다.

Outcome을 제외한 변수들 간의 상관성을 보았을 때, 0.9 이상의 상관관계를 가지는 변수는 없었다. 그러므로 모든 변수를 사용하여 모델링해도 될 것으로 판단된다. 

(2) 이상치를 식별하고 처리하시오. 

우선 describe로 변수들을 살펴보았을 때, 각 변수들의 결측치는 없는 것을 확인하였다. 이상치를 판단하기 위해서는 mean, min, max값을 확인하는 것이 좋다. 평균과 min, 평균과 max 값이 std에 비해 한참 차이가 난다면, 이상치가 있을 가능성이 높다. 
이러한 이상치를 정확히 판단하기 위해서는 아래와 같이 boxplot으로 시각화하여 보는 것이정확하다. 


Age 컬럼에 눈에띄는 이상치(999)가 있어 Age의 중앙값인 29 로 이상치를 대체한다. 

추가적으로 데이터를 살펴 보았을 때, Glucose와 BloodPressure컬럼에는 0이 존재하지 않아야 된다고 판단되었다. 이를 이상치라고 판단하여 이상치를 제외한 중앙값으로 이상치를 대체하였다.

#### 1.1.2. 앞선 두 단계에서 얻은 향후 분석시 고려사항 작성

위에 describe() 함수를 보게 되면, 최솟값과 최댓값 차이가 많이 나는 컬럼이 존재한다. 따라서 선형모델 사용시 scale을 적용할 필요가 보인다.

### 1.2 클래스 불균형을 처리하시오.

#### 1.2.1. 업 샘플링 과정 설명하고 결과 작성

Upsampling은 소수 클래스를 늘려서 다수 클래스 개수에 맞추는 방식으로,  
대표적으로 random으로 소수의 클래스을 선택하여 데이터를 복제하는 RandomOverSampler, 임의의 소수 클래스 주변으로 새로운 데이터를 생성하는 smote 방식이 있다.

#### 1.2.2 언더 샘플링 과정 설명하고 결과 작성

undersampling은 다수 클래스를 감소시켜 소수 클래스 개수에 맞추는 방식으로,  
대표적으로 random으로 다수의 클래스의 데이터를 선택하여 삭제하는 RandomUnderSampler, 서로 다른 클래스가 있을 때 서로 다른 클래스끼리 가장 가까운 데이터들이 토멕링크로 묶여서 토멕링크 중 다수 클래스의 데이터를 제거하는 Tomek link방식이 있다.

#### 1.2.3 둘 중 선택하고 이유 설명

현 데이터가 총 768개로 적은 데이터가 존재한다.. 따라서, undersampling을 선택하게 되면 데이터가 더 적어지기 때문에 오버피팅이 일어날 위험이 더 크다.  
**oversampling을 선택할 것 이다**.

### 1.3. 모델링 하시오

#### 1.3.1. 최소 3개 이상 알고리즘 제시하고 정확도 측면의 모델 1개와 속도 측면의 모델 1개를 꼭 구현(총 2개 이상)

속도 측면에서 logisticregression,정확도 측면에서 svm, 기타로 xgboost를 제시한다

#### 1.3.2. 모델 비교하고 결과 설명

데이터 수가 적기 때문에 hold-out 기법이 아니라 cross-validation 기법으로 성능을 확인하였고, 위에서 말했던거와 같이 oversampling 기법인 smote를 활용하였다.  
따라서, 역시나 logistic이 가장 빨랐으며 svm이 성능이 가장 좋은것을 확인할수있었다.

#### 1.3.3. 속도 개선을 위한 차원 축소 설명하고 수행, 예측 성능과 속도 비교하고 결과 작성

예측 성능은 다소 상승되었고 속도 측면에도 차원축소에 의해 더 빨라졌다.  
향후에 데이터가 크게 늘어난다면 성능 및 속도 측면에서 더 유의미한 차이가 클 것으로 보인다.